### PR TITLE
Add radial tree visualization (fixes #40)

### LINKED ISSUE #40: Add Radial Tree visualization
Add a visualization to display the folder/file heirarchy as a "radial tree" similar to these:

[Radial Tidy Tree](https://observablehq.com/@d3/radial-tree/2)

Image

[Radial Tree](https://vega.github.io/vega/examples/radial-tree-layout/)

Image

Layout as follows

The root folder appears in the centre
Direct children are arranged in a circle around the root folder
Grandchildren are arranged in a wider circle around the outside of that, and so on.

Option to display names - all, folders-only, none
* Any names that are shown are oriented radially - an extended text baseline runs through the centre of the diagram
* Text orientation has "down" towards the bottom of the diagram

Each file is represented by a small disc - metric selection is available for the disc size, fill, and border.

You'll need to research how to lay out the diagram.

### PR DESCRIPTION
**Pull request by theunrepentantgeek**
## Summary

Implements the radial tree visualization requested in #40.

### What's new

- **`internal/radialtree/`** — new package with `RadialNode` types and `Layout()` algorithm: ring-geometry placement, proportional angular sectors by leaf count, metric-scaled disc sizes
- **`internal/render/`** — `RenderRadialPNG()` with three-pass rendering (edges → discs → labels) and radial label rotation keeping text upright on both halves
- **`internal/config/`** — `Radial` config struct, wired into `Config.New()`
- **`cmd/codeviz/`** — `render radial` subcommand with `--disc-size`, `--fill`, `--border`, `--labels`, `--width`, `--height` flags

### Usage

```
codeviz render radial ./myproject -o out.png -d file-size
codeviz render radial ./myproject -o out.png -d file-lines --labels folders --fill file-type
```

### Tests

12 unit tests in `internal/radialtree/layout_test.go` covering root centering, ring placement, disc-size scaling, all three label modes, nested depth ordering, and canvas-size scaling. All 13 test packages green.

Closes #40

### CODE REVIEW

File: internal/render/radialtree.go [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+			dc.SetColor(TextColourFor(fill))
+			dc.DrawStringAnchored(node.Label, sx, sy, 0.5, 0.5)
+		} else {
+			labelRadius := dist + node.DiscRadius + radialLabelGap
+			lx := cx + labelRadius*math.Cos(node.Angle)
+			ly := cy + labelRadius*math.Sin(node.Angle)
+
+			// Normalise angle to [0, 2π) for half-plane check.
+			angle := node.Angle
+			for angle < 0 {
```
Comment (theunrepentantgeek) [inline internal/render/radialtree.go]: **Inefficient angle normalisation loops**

```go
for angle < 0 {
    angle += 2 * math.Pi
}
for angle >= 2*math.Pi {
    angle -= 2 * math.Pi
}
```

These two loops can run O(n) times for angles that are far from `[0, 2π)` (e.g. after many recursive layout passes). The idiomatic, O(1) replacement is:

```go
angle = math.Mod(angle, 2*math.Pi)
if angle < 0 {
    angle += 2 * math.Pi
}
```

Note that `math.Mod` alone is not sufficient — it preserves the sign of the dividend, so negative inputs still yield a negative result, hence the single `if` guard.

File: internal/render/radialtree.go [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+	py := cy + node.Y
+
+	dc.SetColor(radialEdgeColour)
+	dc.SetLineWidth(0.5)
+
+	for _, child := range node.Children {
+		chx := cx + child.X
+		chy := cy + child.Y
+		dc.DrawLine(px, py, chx, chy)
+		dc.Stroke()
```
Comment (theunrepentantgeek) [inline internal/render/radialtree.go]: **`dc.Stroke()` called once per edge — should be batched**

```go
for _, child := range node.Children {
    chx := cx + child.X
    chy := cy + child.Y
    dc.DrawLine(px, py, chx, chy)
    dc.Stroke()          // ← flushes the path after every single line
    drawEdges(dc, child, cx, cy)
}
```

Each `dc.Stroke()` call flushes the current path to the canvas individually. For a tree with many edges this means one GPU/canvas flush per line segment. The more efficient pattern is to batch all `DrawLine` calls and issue a single `Stroke()` after the loop:

```go
for _, child := range node.Children {
    dc.DrawLine(px, py, cx+child.X, cy+child.Y)
}
dc.Stroke()

for _, child := range node.Children {
    drawEdges(dc, child, cx, cy)
}
```

(The recursive calls must stay separate since each level resets colour/width, but the lines for a given node can be batched.)

File: cmd/codeviz/radialtree_cmd.go [hunk 3]
```diff
@@ -0,0 +1,514 @@
...
+	return nil
+}
+
+// resolveLabels converts the string labels flag to a radialtree.LabelMode.
+func (c *RadialCmd) resolveLabels(cfg *config.Radial) radialtree.LabelMode {
+	if lbl := ptrString(cfg.Labels); lbl != "" {
+		return radialtree.LabelMode(lbl)
+	}
+
+	if c.Labels != "" {
```
Comment (theunrepentantgeek) [inline cmd/codeviz/radialtree_cmd.go]: **Dead code branch — `c.Labels` can never be empty here**

```go
func (c *RadialCmd) resolveLabels(cfg *config.Radial) radialtree.LabelMode {
    if lbl := ptrString(cfg.Labels); lbl != "" {
        return radialtree.LabelMode(lbl)   // always taken
    }

    if c.Labels != "" {                    // unreachable
        return radialtree.LabelMode(c.Labels)
    }
    ...
}
```

`Labels` carries `default:"all"` in its Kong struct tag (line 35), so Kong always sets it to a non-empty value. `applyOverrides` (line 176) then unconditionally writes that value into `cfg.Radial.Labels`:

```go
if c.Labels != "" {          // always true — default is "all"
    cfg.Radial.Labels = &c.Labels
}
```

Therefore `ptrString(cfg.Labels)` is always non-empty when `resolveLabels` is called, and the `c.Labels != ""` branch (line 297) is dead code. The fallthrough to `return radialtree.LabelAll` (line 301) is also unreachable. The body can be simplified to:

```go
func (c *RadialCmd) resolveLabels(cfg *config.Radial) radialtree.LabelMode {
    if lbl := ptrString(cfg.Labels); lbl != "" {
        return radialtree.LabelMode(lbl)
    }
    return radialtree.LabelAll
}
```

File: cmd/codeviz/radialtree_cmd.go [hunk 4]
```diff
@@ -0,0 +1,514 @@
...
+			applyRadialFillColours(nodes, root, fillMetric, buckets, numBuckets, fillPalette)
+		}
+	} else {
+		types := collectDistinctTypes(root, fillMetric)
+		mapper := palette.NewCategoricalMapper(types, fillPalette)
+		applyCategoricalRadialFillColours(nodes, root, fillMetric, mapper)
+	}
+}
+
+func (*RadialCmd) applyBorderColours(
```
Comment (theunrepentantgeek) [inline cmd/codeviz/radialtree_cmd.go:344]: **Pointer receiver on a non-mutating method — should be value receiver**

```go
func (*RadialCmd) applyBorderColours(
    nodes *radialtree.RadialNode,
    root *model.Directory,
    cfg *config.Radial,
) (metric.Name, palette.PaletteName) {
```

The receiver is `*RadialCmd` but the method never reads or writes any field of `RadialCmd`. Per the [Go spec and Effective Go](https://go.dev/doc/effective_go#pointers_vs_values), a method should use a value receiver when it does not modify the receiver. This should be:

```go
func (RadialCmd) applyBorderColours(...
```

The same pattern is already used correctly for `resolveFillPalette` on line 256:

```go
func (*RadialCmd) resolveFillPalette(cfg *config.Radial, ...) palette.PaletteName {
```

…which has the same issue. Both should be value receivers for consistency with Go conventions.

File: internal/radialtree/layout_test.go [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+			makeFile("d.go", 100),
+		},
+	}
+
+	node := Layout(root, 800, filesystem.FileSize, LabelAll)
+	g.Expect(node.Children).To(HaveLen(4))
+
+	angles := make(map[float64]bool)
+	for _, child := range node.Children {
+		g.Expect(angles[child.Angle]).To(BeFalse(), "duplicate angle found: %f", child.Angle)
```
Comment (theunrepentantgeek) [inline internal/radialtree/layout_test.go]: **Test quality: `TestLayoutAnglesFullCircle` doesn't verify angle distribution, only uniqueness**

The test name promises "FullCircle" behaviour, but the assertion only checks that no two angles are equal (using a map-uniqueness trick). It does **not** verify that the 4 equal-weight files are spaced ~π/2 apart or that their angles span the full 2π range.

A buggy layout that placed all four nodes in the same quadrant — with four *distinct* but tightly-clustered angles — would pass this test. A meaningful assertion would be something like:

```go
// sort the 4 angles and check each consecutive gap is approximately 2π/4
sort.Float64s(sortedAngles)
expectedGap := 2 * math.Pi / 4
for i := range 3 {
    g.Expect(sortedAngles[i+1] - sortedAngles[i]).To(BeNumerically("~", expectedGap, 0.01))
}
```

Without this, the test provides much weaker coverage than its name implies.

File: cmd/codeviz/radialtree_cmd.go [hunk 6]
```diff
@@ -0,0 +1,514 @@
...
+
+	if c.Border != "" {
+		cfg.Radial.Border = &c.Border
+	}
+
+	if c.BorderPalette != "" {
+		cfg.Radial.BorderPalette = &c.BorderPalette
+	}
+
+	if c.Labels != "" {
```
Comment (theunrepentantgeek) [inline cmd/codeviz/radialtree_cmd.go:192]: `applyOverrides` always overwrites `Labels`, silently ignoring the config file value

`c.Labels` is declared with `default:"all"` in Kong, so it is **never empty** after CLI parsing. The guard on line 176 is therefore always true, meaning `cfg.Radial.Labels` is unconditionally overwritten with the CLI default on every invocation — even when the user hasn't passed `--labels`.

Consequence: a config file entry like `labels: folders` is silently ignored; the user always gets `all`.

The fix is to track whether the flag was explicitly set by the user (e.g. use a `*string` with no Kong default for `Labels` and fall through to the config value when it is nil), so that the config file value is preserved when the user doesn't pass `--labels`.

File: internal/radialtree/layout.go [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+		childStart += childSweep
+	}
+
+	return node
+}
+
+// fileDiscRadius returns the disc pixel radius for f, scaled by the disc metric.
+func fileDiscRadius(f *model.File, discMetric metric.Name, dp discParams) float64 {
+	val := fileMetricValue(f, discMetric)
+	if val <= 0 {
```
Comment (theunrepentantgeek) [inline internal/radialtree/layout.go:162]: **Missing test: `fileDiscRadius` zero-value metric path (val ≤ 0)**

When a file has no value for the disc metric — e.g. it was scanned but the metric was never recorded — `fileMetricValue` returns 0 and the `val <= 0` branch returns `dp.fileMin`. This floor is important: it prevents zero-radius discs from becoming invisible.

None of the 12 tests exercise this path. `TestLayoutDiscSizeScalesWithMetric` uses sizes 100 and 1000; `TestLayoutSingleFile` uses size 100. Every test file has a valid, positive metric.

A simple addition to the test suite would be:

```go
func TestLayoutZeroMetricFileGetsMinDisc(t *testing.T) {
    t.Parallel()
    g := NewGomegaWithT(t)

    // makeFile with size 0 → fileMetricValue returns 0 → should get fileMin disc
    root := &model.Directory{
        Name:  "root",
        Files: []*model.File{makeFile("empty.go", 0)},
    }
    node := Layout(root, 800, filesystem.FileSize, LabelAll)
    g.Expect(node.Children[0].DiscRadius).To(BeNumerically("==", minFileDisc))
}
```

Without this, a regression that accidentally returned 0 (or panicked on a zero-valued file) would go undetected.

File: cmd/codeviz/radialtree_cmd.go (hunk 3 above)
Comment (theunrepentantgeek) [inline cmd/codeviz/radialtree_cmd.go]: **`resolveLabels` has two unreachable branches**

Because `applyOverrides` always sets `cfg.Radial.Labels` (the `c.Labels != ""` guard is always true due to Kong's `default:"all"`), the first branch of `resolveLabels` always fires and returns. The second branch (`if c.Labels != ""`) and the final `return radialtree.LabelAll` are therefore dead code.

This isn't just cosmetic — it indicates that the intended control flow (config file → CLI flag → built-in default) is not actually implemented. Cleaning up these branches will make the intent clear once the `applyOverrides` bug is fixed.

File: internal/radialtree/layout.go [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+		return dp.fileMin
+	}
+
+	if dp.useEqual {
```
Comment (theunrepentantgeek) [inline internal/radialtree/layout.go:166]: **Missing test: `fileDiscRadius` uniform-metric (`useEqual`) branch**

When every file in the tree has the **same** non-zero metric value, `buildDiscParams` sets `useEqual = true` (line 69) and `fileDiscRadius` returns the midpoint `(fileMin + fileMax) / 2`. This branch exists precisely to avoid a divide-by-zero in the scaled formula at line 156.

No test covers it. `TestLayoutDiscSizeScalesWithMetric` uses sizes 100 and 1000 (different), so it always takes the normal scaling path. A tree where every file has size 500 would never hit this branch in the current test suite.

A targeted test:

```go
func TestLayoutUniformMetricGetsMidpointDisc(t *testing.T) {
    t.Parallel()
    g := NewGomegaWithT(t)

    // All files have the same size → useEqual = true → disc = (fileMin+fileMax)/2
    root := &model.Directory{
        Name: "root",
        Files: []*model.File{
            makeFile("a.go", 500),
            makeFile("b.go", 500),
            makeFile("c.go", 500),
        },
    }
    node := Layout(root, 800, filesystem.FileSize, LabelAll)
    for _, child := range node.Children {
        g.Expect(child.DiscRadius).To(BeNumerically(">", minFileDisc))
    }
    // All discs should be identical (midpoint)
    g.Expect(node.Children[0].DiscRadius).To(BeNumerically("==", node.Children[1].DiscRadius))
}
```

Without this, a divide-by-zero regression in the equal-metric path would silently survive the test suite.

File: internal/render/radialtree.go [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+	radialDefaultBorder   = color.RGBA{R: 0x33, G: 0x33, B: 0x33, A: 0xFF}
+	radialEdgeColour      = color.RGBA{R: 0x99, G: 0x99, B: 0x99, A: 0xFF}
+)
+
+const radialLabelGap = 4.0
+
+// RenderRadialPNG renders the radial tree layout to a PNG file at the given path.
+// Drawing is done in three passes — edges, then discs, then labels — to ensure
+// correct z-ordering across the entire tree.
+func RenderRadialPNG(root radialtree.RadialNode, canvasSize int, outputPath string) error {
```
Comment (theunrepentantgeek) [inline internal/render/radialtree.go]: **`RenderRadialPNG` accepts `root` by value, copying the entire tree**

`RadialNode` contains a `Children []RadialNode` slice, so every node owns its subtree. Passing the root by value copies the entire tree — for a repository with thousands of files that's a non-trivial allocation before the first pixel is drawn.

The three internal drawing helpers already accept `radialtree.RadialNode` by value when recursing, which is fine for the recursive calls (they receive individual nodes). The API boundary at `RenderRadialPNG`, however, is the entry point for the full tree and should take a pointer:

```go
func RenderRadialPNG(root *radialtree.RadialNode, canvasSize int, outputPath string) error
```

The call site in `radialtree_cmd.go` passes `nodes` (already a `RadialNode` value), so it would become `&nodes`.

File: cmd/codeviz/radialtree_cmd.go [hunk 10]
```diff
@@ -0,0 +1,514 @@
...
+	dir *model.Directory,
+	m metric.Name,
+	buckets metric.BucketBoundaries,
+	numBuckets int,
+	p palette.ColourPalette,
+) {
+	fileIdx := 0
+	dirIdx := 0
+
+	for i := range node.Children {
```
Comment (theunrepentantgeek) [inline cmd/codeviz/radialtree_cmd.go:444]: **Colour-application functions are silently coupled to `layoutDir`'s file-first child ordering**

`applyRadialFillColours` (and its border/categorical siblings) correctly assign colours today because `layoutDir` appends files before subdirectories. The logic relies on this invariant without documenting or enforcing it: if `layoutDir` ever reorders children (e.g. to sort alphabetically across both files and dirs), colours would silently misassign — files would get directory colours and vice versa — with no compile-time or runtime error.

Consider one of:
1. **Document the contract** — add a comment here and in `layoutDir` cross-referencing the ordering dependency.
2. **Zip layout and model together** — have `Layout` return paired `(RadialNode, *model.Directory)` or store a back-reference in `RadialNode`, so colour application doesn't need to walk a separate model tree in lock-step.
3. **Keep the dependency but make it structural** — store file children and dir children in separate slices on `RadialNode`, eliminating the `IsDirectory` dispatch entirely.

**Comment by theunrepentantgeek** [comment]
**Coverage gap: `RenderRadialPNG` has zero tests**

`internal/render/radialtree.go` ships with no test file. The analogous treemap renderer (`renderer.go`) has a thorough test file (`renderer_test.go`) with multiple smoke tests that follow a clear, reusable pattern:

```go
func TestRenderRadialPNG_Smoke(t *testing.T) {
    t.Parallel()
    g := NewGomegaWithT(t)

    root := &model.Directory{
        Name: "root",
        Files: []*model.File{
            makeFile("a.go", "go", 100),
            makeFile("b.go", "go", 500),
        },
    }
    node := radialtree.Layout(root, 800, filesystem.FileSize, radialtree.LabelAll)
    out := filepath.Join(t.TempDir(), "radial.png")
    err := RenderRadialPNG(node, 800, out)
    g.Expect(err).NotTo(HaveOccurred())

    info, err := os.Stat(out)
    g.Expect(err).NotTo(HaveOccurred())
    g.Expect(info.Size()).To(BeNumerically(">", 0))
}
```

Without any tests, the three rendering passes (`drawEdges`, `drawDiscs`, `drawLabels`), the two label-orientation branches in `drawLabels` (right-half vs left-half of canvas), and the error path from `dc.SavePNG` are completely untested. A crash or silent black-square output would not be caught by CI.

File: internal/config/radialtree.go [hunk 11]
```diff
@@ -0,0 +1,11 @@
+package config
+
+// Radial holds persistent configuration for radial tree visualizations.
```
Comment (theunrepentantgeek) [inline internal/config/radialtree.go:3]: **`Radial` config struct is a near-duplicate of `Treemap` — consider a shared base**

`Treemap` has `Fill`, `FillPalette`, `Border`, `BorderPalette`. `Radial` has all four of those plus `Labels`. The shared four fields are byte-for-byte identical including YAML/JSON tags.

As more render modes are added, this duplication will compound. A lightweight fix is to introduce a shared embedded struct for the common palette fields:

```go
// ColourConfig holds the palette fields shared by all render modes.
type ColourConfig struct {
    Fill          *string `yaml:"fill,omitempty"          json:"fill,omitempty"`
    FillPalette   *string `yaml:"fillPalette,omitempty"   json:"fillPalette,omitempty"`
    Border        *string `yaml:"border,omitempty"        json:"border,omitempty"`
    BorderPalette *string `yaml:"borderPalette,omitempty" json:"borderPalette,omitempty"`
}

type Treemap struct{ ColourConfig }
type Radial  struct{ ColourConfig; Labels *string `yaml:"labels,omitempty" json:"labels,omitempty"` }
```

Embedding preserves identical YAML/JSON serialization and lets shared helpers accept `*ColourConfig` instead of duplicating the resolver logic.

File: internal/radialtree/layout.go [hunk 12]
```diff
@@ -0,0 +1,228 @@
...
+
+	if v > hi {
+		return hi
+	}
+
+	return v
+}
+
+// computeLeafCount returns the total number of file leaves under dir.
+// Empty directories are treated as having 1 leaf to avoid zero-division in sector math.
```
Comment (theunrepentantgeek) [inline internal/radialtree/layout.go]: **Misleading doc comment on `computeLeafCount` + nested-empty-dir not tested**

The function comment says:
> Empty directories are treated as having 1 leaf to avoid zero-division in sector math.

But `computeLeafCount` itself returns **0** for an empty directory — it never substitutes 1. The invariant is enforced by the two guard blocks in `layoutDir` (lines 102–104 and 131–133). A future reader relying on the comment's promise could remove those guards thinking the function already handles it.

Additionally, the guard is only reached via `TestLayoutEmptyDirectory`, which tests a completely empty root (no subdirs). It does **not** test the case of `root → emptySubDir → emptySubSubDir`, where `computeLeafCount(root)` would recurse and still return 0. That scenario exercises the `parentLeafCount = 1` guard but also the subdir loop's `childLeafCount = 1` guard simultaneously.

Two suggested fixes:
1. Correct the comment to say the function returns the raw count (possibly 0) and that callers must guard against division by zero.
2. Add a test: a root with one subdir that itself contains only an empty subdir — assert it doesn't panic and the layout produces nodes.

File: internal/render/radialtree.go (hunk 1 above)
Comment (theunrepentantgeek) [reply internal/render/radialtree.go]: Will fix.

File: internal/render/radialtree.go (hunk 2 above)
Comment (theunrepentantgeek) [reply internal/render/radialtree.go]: Will fix.

File: cmd/codeviz/radialtree_cmd.go (hunk 3 above)
Comment (theunrepentantgeek) [reply cmd/codeviz/radialtree_cmd.go]: The Kong based CLI won't always be the only client, so it will be possible for c.Labels to be empty.

File: cmd/codeviz/radialtree_cmd.go (hunk 4 above)
Comment (theunrepentantgeek) [reply cmd/codeviz/radialtree_cmd.go:344]: This projects convention is that all method receivers for a type must be consistent - all pointers or all values - unless there's a very compelling reason for a special case.

File: internal/radialtree/layout_test.go (hunk 5 above)
Comment (theunrepentantgeek) [reply internal/radialtree/layout_test.go]: Please improve the test coverage.

File: cmd/codeviz/radialtree_cmd.go (hunk 6 above)
Comment (theunrepentantgeek) [reply cmd/codeviz/radialtree_cmd.go:192]: All Kong structs should use pointers for exactly this reason - so we can tell whether or not a value has been applied. Defaults should be handled by config.New(). 

Check whether Kong defaults are _applied_ by Kong, or just serve as documentation. If they're applied, remove them in favour of relying on config.New() and update the help text to reflect the default.

File: internal/radialtree/layout.go (hunk 7 above)
Comment (theunrepentantgeek) [reply internal/radialtree/layout.go:162]: Please add a test.

File: cmd/codeviz/radialtree_cmd.go (hunk 3 above)
Comment (theunrepentantgeek) [reply cmd/codeviz/radialtree_cmd.go]: Please review this after the Kong struct changes are complete and let me know if the code is still unreachable.

File: internal/radialtree/layout.go (hunk 8 above)
Comment (theunrepentantgeek) [reply internal/radialtree/layout.go:166]: Add a test please.

File: internal/render/radialtree.go (hunk 9 above)
Comment (theunrepentantgeek) [reply internal/render/radialtree.go]: Please fix.

File: cmd/codeviz/radialtree_cmd.go (hunk 10 above)
Comment (theunrepentantgeek) [reply cmd/codeviz/radialtree_cmd.go:444]: For now, just document the contract in the code.

File: internal/config/radialtree.go (hunk 11 above)
Comment (theunrepentantgeek) [reply internal/config/radialtree.go:3]: They are accidentally similar and shouldn't be coupled because they have different reasons to change.

File: internal/radialtree/layout.go (hunk 12 above)
Comment (theunrepentantgeek) [reply internal/radialtree/layout.go]: Please apply both fixes.

**Comment by theunrepentantgeek** [comment]
> internal/render/radialtree.go ships with no test file. 

Please add a test suite for this.

**Comment by theunrepentantgeek** [comment]
🪲 Many node labels are missing from the final image (perhaps because they're drawn in white?).

🪲The fixed size dots used for directories are disproportionally large; they should be similar in size to "small" file nodes.

🪲 When there are lots of leaf nodes in the radial, they're overlapping. If a circle is crowded, we need to increase the radius so that items on the circumference are spaced out and don't overlap. 

🪲 Large file nodes can obscure smaller ones; draw the larger ones first so they appear behind smaller ones
